google-cloud-firestore/src/main/java/com/google/cloud/firestore/QuerySnapshot.java

```diff
@@ -16,48 +16,137 @@
 
 package com.google.cloud.firestore;
 
-import com.google.common.base.Preconditions;
+import com.google.cloud.firestore.DocumentChange.Type;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
 import javax.annotation.Nonnull;
-import javax.annotation.Nullable;
 import org.threeten.bp.Instant;
 
 /**
  * A QuerySnapshot contains the results of a query. It can contain zero or more DocumentSnapshot
  * objects.
  */
-public final class QuerySnapshot implements Iterable<QueryDocumentSnapshot> {
+public abstract class QuerySnapshot implements Iterable<QueryDocumentSnapshot> {
 
   private final Query query;
-  private final DocumentSet documentSet;
-  @Nullable private volatile List<QueryDocumentSnapshot> documentSnapshots;
   private final Instant readTime;
-  private final List<DocumentChange> documentChanges;
 
-  QuerySnapshot(
-      Query query,
-      Instant readTime,
-      List<QueryDocumentSnapshot> results,
-      List<DocumentChange> documentChanges) {
+  private QuerySnapshot(Query query, Instant readTime) {
     this.query = query;
-    this.documentSet = null;
-    this.documentSnapshots = results;
     this.readTime = readTime;
-    this.documentChanges = documentChanges;
   }
 
-  QuerySnapshot(
-      Query query, Instant readTime, DocumentSet results, List<DocumentChange> documentChanges) {
-    this.query = query;
-    this.documentSet = results;
-    this.readTime = readTime;
-    this.documentChanges = documentChanges;
+  /** Creates a new QuerySnapshot representing the results of a Query with added documents. */
+  public static QuerySnapshot withDocuments(
+      final Query query, Instant readTime, final List<QueryDocumentSnapshot> documents) {
+    return new QuerySnapshot(query, readTime) {
+      volatile List<DocumentChange> documentChanges;
+
+      @Nonnull
+      @Override
+      public List<QueryDocumentSnapshot> getDocuments() {
+        return Collections.unmodifiableList(documents);
+      }
+
+      @Nonnull
+      @Override
+      public List<DocumentChange> getDocumentChanges() {
```

Comment:
If it helps, I recently updated auto-value dependency. The `@Memoized` annotation should work now and might help with this.

Reply:
Oh cool, but I still have some bad news: It looks like if I want to use this, I would have to separate out the two cases into their own AutoValue class (since the data type that they reference is different). I tried to just make QuerySnapshot itself AutoValue, but the computations itself then have to happen inside this class as well.